### PR TITLE
Wallet and credit test

### PR DESCRIPTION
**Pull request by NataliaHivelocity**

### CODE REVIEW

File: apps/hub_payments/lib/hub_payments_web/controllers/api/v1/credit_card_controller.ex [hunk 1]
```diff
@@ -4,10 +4,40 @@ defmodule HubPaymentsWeb.Api.V1.CreditCardController do
...
   alias HubPayments.Wallets.Wallet
 
+  def index(conn, %{"wallet_uuid" => wallet_uuid}) do
+    credit_cards = Wallets.list_credit_cards(%{wallet_uuid: wallet_uuid})
+    render(conn, "index.json", credit_cards: credit_cards)
+  end
+
+  def show(conn, %{"wallet_uuid" => wallet_uuid, "credit_card_uuid" => credit_card_uuid}) do
+    with credit_card =
+           Wallets.get_credit_card(%{uuid: credit_card_uuid, wallet_uuid: wallet_uuid}) do
```
Comment (PatOkochi) [inline apps/hub_payments/lib/hub_payments_web/controllers/api/v1/credit_card_controller.ex:14]: change the = for an <-
```
credit_card <-
           Wallets.get_credit_card(%{uuid: credit_card_uuid, wallet_uuid: wallet_uuid}) 
```

File: apps/hub_payments/lib/hub_payments_web/controllers/api/v1/credit_card_controller.ex [hunk 2]
```diff
@@ -4,10 +4,40 @@ defmodule HubPaymentsWeb.Api.V1.CreditCardController do
...
     end
   end
+
+  def update(conn, %{
+        "wallet_uuid" => wallet_uuid,
+        "credit_card_uuid" => credit_card_uuid,
+        "credit_card_params" => credit_card_params
+      }) do
+    with credit_card =
+           Wallets.get_credit_card(%{uuid: credit_card_uuid, wallet_uuid: wallet_uuid}),
```
Comment (PatOkochi) [inline apps/hub_payments/lib/hub_payments_web/controllers/api/v1/credit_card_controller.ex:32]: same here
```
%CreditCard{} = credit_card <-
           Wallets.get_credit_card(%{uuid: credit_card_uuid, wallet_uuid: wallet_uuid}) 
```

File: apps/hub_payments/test/hub_payments_web/controllers/api/v1/credit_card_controller_test.exs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+      response =
+        put(conn, "/api/v1/wallets/#{wallet.uuid}/credit_cards/#{credit_card.uuid}", %{
+          "wallet_uuid" => wallet.uuid,
+          "credit_card_uuid" => credit_card.uuid,
+          "credit_card_params" => @invalid_attrs
+        })
+        |> json_response(400)
+
+      assert response["error"] != %{}
```
Comment (PatOkochi) [inline apps/hub_payments/test/hub_payments_web/controllers/api/v1/credit_card_controller_test.exs:142]: May be good to put the error.

`assert response["error"] == %{"exp_month" => ["can't be blank"], "exp_year" => ["can't be blank"]}`

File: apps/hub_payments/test/hub_payments_web/controllers/api/v1/wallet_controller_test.exs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  #     assert_error_sent 404, fn ->
+  #       get(conn, Routes.wallet_path(conn, :show, wallet))
+  #     end
+  #   end
+  # end
+
+  # defp create_wallet(_) do
+  #   wallet = fixture(:wallet)
+  #   %{wallet: wallet}
+  # end
```
Comment (PatOkochi) [inline apps/hub_payments/test/hub_payments_web/controllers/api/v1/wallet_controller_test.exs:96]: Should get rid of this if we are not gonna use it

**Review by PatOkochi — COMMENTED** [review]
Just a few changes and its fine 👍

File: apps/hub_payments/lib/hub_payments_web/controllers/api/v1/credit_card_controller.ex [hunk 5]
```diff
@@ -4,10 +4,40 @@ defmodule HubPaymentsWeb.Api.V1.CreditCardController do
...
+    credit_cards = Wallets.list_credit_cards(%{wallet_uuid: wallet_uuid})
+    render(conn, "index.json", credit_cards: credit_cards)
+  end
+
+  def show(conn, %{"wallet_uuid" => wallet_uuid, "credit_card_uuid" => credit_card_uuid}) do
+    with credit_card <-
+           Wallets.get_credit_card(%{uuid: credit_card_uuid, wallet_uuid: wallet_uuid}) do
+      render(conn, "show.json", %{credit_card: credit_card})
+    end
+  end
```
Comment (PatOkochi) [inline apps/hub_payments/lib/hub_payments_web/controllers/api/v1/credit_card_controller.ex:17]: We may have to change this.

```
def show(conn, %{"wallet_uuid" => wallet_uuid, "credit_card_uuid" => credit_card_uuid}) do
    with %CreditCard{} = credit_card <-
           Wallets.get_credit_card(%{uuid: credit_card_uuid, wallet_uuid: wallet_uuid}) do
      render(conn, "show.json", %{credit_card: credit_card})
    else
      nil -> {:error, "Credit card not found"}
    end
  end
```

File: apps/hub_payments/lib/hub_payments_web/controllers/api/v1/credit_card_controller.ex [hunk 6]
```diff
@@ -4,10 +4,40 @@ defmodule HubPaymentsWeb.Api.V1.CreditCardController do
   alias HubPayments.Wallets
   alias HubPayments.Wallets.Wallet
```
Comment (PatOkochi) [inline apps/hub_payments/lib/hub_payments_web/controllers/api/v1/credit_card_controller.ex:5]: Gonna need to alias CreditCard too.

`alias HubPayments.Wallets.{CreditCard, Wallet}`

File: apps/hub_payments/lib/hub_payments_web/controllers/api/v1/credit_card_controller.ex [hunk 7]
```diff
@@ -4,10 +4,40 @@ defmodule HubPaymentsWeb.Api.V1.CreditCardController do
...
+           Wallets.get_credit_card(%{uuid: credit_card_uuid, wallet_uuid: wallet_uuid}),
+         {:ok, credit_card} <- Wallets.update_credit_card(credit_card, credit_card_params) do
+      render(conn, "show.json", %{credit_card: credit_card})
+    end
+  end
+
+  def delete(conn, %{"credit_card" => credit_card}) do
+    with {:ok, _} <- Wallets.delete_credit_card(credit_card) do
+      send_resp(conn, :no_content, "")
+    end
```
Comment (PatOkochi) [inline apps/hub_payments/lib/hub_payments_web/controllers/api/v1/credit_card_controller.ex:41]: Change this to receive credit_card_uuid and wallet_uuid to delete the credit card.

File: apps/hub_payments/lib/hub_payments_web/controllers/api/v1/credit_card_controller.ex [hunk 8]
```diff
@@ -4,10 +4,40 @@ defmodule HubPaymentsWeb.Api.V1.CreditCardController do
...
+        "wallet_uuid" => wallet_uuid,
+        "credit_card_uuid" => credit_card_uuid,
+        "credit_card_params" => credit_card_params
+      }) do
+    with credit_card <-
+           Wallets.get_credit_card(%{uuid: credit_card_uuid, wallet_uuid: wallet_uuid}),
+         {:ok, credit_card} <- Wallets.update_credit_card(credit_card, credit_card_params) do
+      render(conn, "show.json", %{credit_card: credit_card})
+    end
+  end
```
Comment (PatOkochi) [inline apps/hub_payments/lib/hub_payments_web/controllers/api/v1/credit_card_controller.ex:36]:
```
def update(conn, %{
        "wallet_uuid" => wallet_uuid,
        "credit_card_uuid" => credit_card_uuid,
        "credit_card_params" => credit_card_params
      }) do
    with %CreditCard{} = credit_card <-
           Wallets.get_credit_card(%{uuid: credit_card_uuid, wallet_uuid: wallet_uuid}),
         {:ok, credit_card} <- Wallets.update_credit_card(credit_card, credit_card_params) do
      render(conn, "show.json", %{credit_card: credit_card})
    else
      nil -> {:error, "Credit card not found"}
    end
    end
  end
```

**Review by PatOkochi — COMMENTED** [review]
Few more changes

File: apps/hub_payments/lib/hub_payments_web/controllers/api/v1/credit_card_controller.ex (hunk 5 above)
Comment (PatOkochi) [reply apps/hub_payments/lib/hub_payments_web/controllers/api/v1/credit_card_controller.ex:17]: Also need tests for this, when credit_card_uuid is nil

**Review by PatOkochi — APPROVED** [review]
Everything is working fine. Well done 😄

**Review by ErinHivelociy — APPROVED** [review]
